### ISSUE TITLE
Slowly converging 1-R and T and unexpected index averaging at PML boundaries

### ISSUE THREAD

**Issue by soamaven** [stevengj/meep#467]
## Issue
- In a 3D simulation, If I place a flux monitor on either side of a lossless planar interface, and on either side of a plane-wave gaussian source, the 1-R and T spectrum values agree well, within 1e-7 for a resolution of ~40px/um (spectrum not shown).

  When I add a periodic scatterer to the surface of the planar interface (in the code below, a crossed grating), the 1-R and T values disagree to 1E-2, and much greater in other parts of the spectrum (see low-res plot). The result does seem to get better with resolution, I've tried 20, 40, 60, 100, and at 100, they agree within 1E-3. Is this probably because of the now scattered fields, with new Ey and Ez components that weren't present in the planar case, introducing more floating points into the linear algebra, reducing precision?
  The time-stepping is taking the longest, the DFTs a small fraction of the total time. The best convergence I've seen is for a resolution of 100, which took 2:10 on 4 i7-8550U cores. I'd hope it was faster :(
  Which spectra, R or T, should I trust?

- Also, the permittivity distribution ![image](https://user-images.githubusercontent.com/14949210/44238638-723ce980-a16a-11e8-9119-fd62307906f9.png) at the bottom of the simulation shows index averaging. I have specified index averaging to be turned off. This probably isn't affecting the simulation results because the fields will have decayed significantly, but it is unintended. I suppose it could exacerbate reflections for shorter PML distances.
At the substrate/air interface, I expect this as [a result of off-yee-grid interpolation at interfaces](https://meep.readthedocs.io/en/latest/FAQ/#why-doesnt-turning-off-subpixel-averaging-work) resulting in the index seen, but why would this occur at the PML edges? Is this because of periodic boundaries in z, the top and bottom yee-grids see each other?  Am I improperly implementing the boundary conditions?

## Script

```
import meep as mp
import numpy as np
import matplotlib.pyplot as plt

nSiO2 = 1.55
SiO2 = mp.Medium(index=nSiO2)
low_index_SiO2 = mp.Medium(index=1.42)

# conversion factor for eV to 1/um
ev_um_scale = 1 / 1.23984193

lambda_min = 0.4  # minimum source wavelength
lambda_max = 1.2  # maximum source wavelength
fmin = 1 / lambda_max
fmax = 1 / lambda_min
fcen = 0.5 * (fmin + fmax)
df = fmax - fmin
nfreq = len(np.arange(lambda_min * 1000, lambda_max * 1000 + 1, 10))

cube_side = .36 * 2  # cube side length
cube_h = .65  # cube height
a = cube_side * 1.5  # lattice constant
Np = 1  # number of lattice periods in simulation

dsub = lambda_max / nSiO2 * 1  # substrate thickness
dpml = lambda_max * 1  # PML thickness
dair = (lambda_max + cube_h) * 1  # air gap thickness

sx = Np * a
sy = Np * a
sz = dpml + dair + dsub + dpml + cube_h

resolution = 100
print(f"Resolution: {resolution}")
print(f"Mesh Size: {sx // (1 / resolution)}x{sy // (1 / resolution)}x{sz // (1 / resolution)}")

cell_size = mp.Vector3(sx, sy, sz)
boundary_layers = [mp.PML(dpml, direction=mp.Z,)]

sub_center = -0.5 * sz + 0.5 * (dsub + dpml)
cube_center = (sub_center + (dsub + dpml) / 2) + 0.5 * cube_h
trans_z = (sub_center + (dsub + dpml) / 2) - (dsub / 2)
source_z = cube_center + 0.5 * cube_h + 0.5 * dair
refl_z = source_z + 0.25 * dair

geometry = [
    mp.Block(material=SiO2, size=mp.Vector3(mp.inf, mp.inf, dsub + dpml),
             center=mp.Vector3(0, 0, sub_center)),
]

for nx in range(Np):
    for ny in range(Np):
        cx = -0.5 * sx + (nx + 0.5) * a
        cy = -0.5 * sy + (ny + 0.5) * a
        geometry.append(mp.Block(material=low_index_SiO2, size=mp.Vector3(cube_side, cube_side, cube_h),
                                 center=mp.Vector3(cx, cy, cube_center)))

for shape in geometry:
    print(f"Geo center: {shape.center}")

k_point = mp.Vector3(0, 0, 0)

sources = [mp.Source(mp.GaussianSource(fcen, fwidth=df, cutoff=10), component=mp.Ex,
                     center=mp.Vector3(0, 0, source_z), size=mp.Vector3(sx, sy, 0))]

symmetries = [mp.Symmetry(direction=mp.X, phase=-1),
              mp.Symmetry(direction=mp.Y, phase=1)
              ]
# symmetries = []

# Normalization Simulation; Empty space
sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    boundary_layers=boundary_layers,
                    geometry=[],
                    dimensions=3,
                    k_point=k_point,
                    sources=sources,
                    symmetries=symmetries,
                    eps_averaging=False,
                    verbose=True,
                    )

refl_fr = mp.FluxRegion(center=mp.Vector3(0, 0, refl_z), size=mp.Vector3(sx, sy, 0))
refl_fm = sim.add_flux(fcen, df, nfreq, refl_fr)
trans_fr = mp.FluxRegion(center=mp.Vector3(0, 0, trans_z), size=mp.Vector3(sx, sy, 0))
trans_fm = sim.add_flux(fcen, df, nfreq, trans_fr)

print(f"Geometry z Extent:{geometry[0].center.z + 0.25 * sz} , {geometry[0].center.z - 0.25 * sz}")
print(f"Geometry x Extent:{geometry[0].center.x + 0.25 * sx} , {geometry[0].center.x - 0.25 * sx}")
print(f'Geometry y Extent:{geometry[0].center.y + 0.25 * sy} , {geometry[0].center.y - 0.25 * sy}')
print(f"Source z:{sources[0].center.z}")
print(f"R z:{refl_fr.center.z}")
print(f"T z:{trans_fr.center.z}")

t_end = 200 / fmin
# sim.run(until_after_sources=mp.stop_when_fields_decayed(15, mp.Ex, mp.Vector3(0, 0, cube_center), 1e-3))
sim.run(until=t_end)

refl_norm_flux_dat = sim.get_flux_data(refl_fm)
trans_norm_flux_dat = sim.get_flux_data(trans_fm)

# Store the flux data for normalization later
refl_norm_flux = np.array(mp.get_fluxes(refl_fm))
trans_norm_flux = np.array(mp.get_fluxes(trans_fm))

sim.reset_meep()
# Crossed grating Simulation
sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    boundary_layers=boundary_layers,
                    geometry=geometry,
                    dimensions=3,
                    k_point=k_point,
                    sources=sources,
                    symmetries=symmetries,
                    eps_averaging=False,
                    verbose=True,
                    )

refl_fm = sim.add_flux(fcen, df, nfreq, refl_fr)
trans_fm = sim.add_flux(fcen, df, nfreq, trans_fr)

source_fr = mp.FluxRegion(center=mp.Vector3(0, 0, source_z), size=mp.Vector3(sx, sy, 0))
source_fm = sim.add_flux(fcen, df, nfreq, source_fr)


sim.load_minus_flux_data(refl_fm, refl_norm_flux_dat)
# sim.load_minus_flux_data(trans, trans_norm_flux_dat)

# sim.run(until_after_sources=mp.stop_when_fields_decayed(15, mp.Ex, mp.Vector3(0, 0, cube_center), 1e-3))
sim.run(until=t_end)

# Get the fluxes from the crossed grating simulation to normalize
refl_cube_flux = np.array(mp.get_fluxes(refl_fm))
trans_cube_flux = np.array(mp.get_fluxes(trans_fm))

flux_freqs = mp.get_flux_freqs(refl_fm)
# sim.display_fluxes(refl_fm, trans_fm)

# Normalize to the flux that is input into a free space simulation (The transmitted flux recorded in just free space)
wl = 1 / np.array(flux_freqs)
Rs = refl_cube_flux / trans_norm_flux
Ts = trans_cube_flux / trans_norm_flux

f, (ax0, ax1, ax2) = plt.subplots(1, 3, figsize=(16, 12))
ax0.plot(wl, 1 + Rs, 'bo-', label='1 - reflectance')
ax0.plot(wl, Ts, 'rx-', label='transmittance')
ax0.plot(wl, np.abs(Ts) + np.abs(Rs), 'k+-', label='T + R')
ax0.plot(wl, -1 * FSRs, 'go-', label='FreeSpace R')
ax0.set_ylim((.85, 1.15))
ax0.set(xlabel="wavelength (μm)")
ax0.legend(loc="upper right")
ax0.set_aspect("auto", )

eps_data = sim.get_array(center=mp.Vector3(), size=cell_size, component=mp.Dielectric)
x = np.linspace(-0.5 * sx, 0.5 * sx, eps_data.shape[0])
y = np.linspace(-0.5 * sy, 0.5 * sy, eps_data.shape[1])
z = np.linspace(-0.5 * sz, 0.5 * sz, eps_data.shape[2])
cube_center_z_index = np.where(z <= cube_center)[0][-1]
# xz cross-section through y = 0
ax1.pcolormesh(x, z, eps_data[:, eps_data.shape[1] // 2, :].transpose(), cmap='viridis')
ax1.hlines(y=sources[0].center.z, xmin=-0.5 * sx, xmax=0.5 * sx, colors='r')
ax1.hlines(y=refl_fr.center.z, xmin=-0.5 * sx, xmax=0.5 * sx, colors='b')
ax1.hlines(y=trans_fr.center.z, xmin=-0.5 * sx, xmax=0.5 * sx, colors='g')
ax1.hlines(y=0.5 * sz - dpml, xmin=-0.5 * sx, xmax=0.5 * sx, colors='k')
ax1.hlines(y=-0.5 * sz + dpml, xmin=-0.5 * sx, xmax=0.5 * sx, colors='k')
ax1.set_aspect('equal', )
# xy cross-section through cubes
ax2.pcolormesh(x, y, eps_data[:, :, cube_center_z_index].transpose(),
               cmap='viridis')
ax2.set_aspect('equal', )
plt.show()
```
### Non-equal 1-R and T spectra at high resolution (100px/um)
![image](https://user-images.githubusercontent.com/14949210/44247146-2d7b7780-a197-11e8-9e51-010ba92d8183.png)
Starting to converge, but not near the precision of the planar interface. Simulations at such high resolution or greater may be time prohibitive.
### Low resolution (20 px/um) dielectric distribution to clearly show unexpected index averaging at boundaries
![image](https://user-images.githubusercontent.com/14949210/44244558-5b59bf80-a189-11e8-8eff-bb16d51ea378.png)
## Other notes (hopefully helpful to someone in the future, if not to this issue):
- I've read [the FAQ](https://meep.readthedocs.io/en/latest/FAQ/#why-are-there-strange-peaks-in-my-reflectancetransmittance-spectrum-when-modeling-planar-or-periodic-structures) about "spikey" spectra, and this doesn't seem to be my situation, unless I am applying the boundary condition improperly.
- I was initially having issues with `sim.run(until_after_sources=mp.stop_when_fields_decayed(50, mp.Ex, mp.Vector3(0, 0, cube_center), 1e-3))` and the simulation would not converge. Perhaps because I was checking a symmetry point in the center, there were issues.There should be a first order mode with low Q that would decay over time, but the simulation would not end. Things work as expected for a point off-symmetry. Perhaps this is a bug, or should be mentioned to not look for field decay along symmetry intersections. I can open another issue where this is reproduced.
-  Simulations seem to run faster when geometric objects are integer multiples of the pixels/wavelength. 
- I compiled from source using `gnu` compilers with `mkl`, and passed all tests, `cpp` and `python`. The compile bash script was just sent out to the email list, in case it is relevant, or can help someone else.
- Using `python 3.6`
- I am hoping to measure normalized differences to 3 or more decimals, any advice here would be helpful. I am coming from Lumerical, caught the FOSS bug, and want to see if MEEP will fit my needs. I want to do many simulations for optimization, and was hoping that I could get comparable precision and faster simulations by compiling for the target machine.
  - Results get better when index averaging is turned on, but still not better than 1E-3

**Comment by oskooi** [comment]
In flux calculations which typically involve pulsed sources, you should be using the termination condition `stop_when_fields_decayed` (this is described in the [documentation](https://meep.readthedocs.io/en/latest/Introduction/#transmittancereflectance-spectra)). Using a fixed runtime as you are doing here, particularly for the "real" run involving the scatterer, may likely result in terminating the simulation prematurely (i.e., such that the fields produced by the source have not sufficiently decayed away). You can check this by gradually increasing the runtime `t_end` and verifying that in the limit that `t_end` approaches infinity the results have converged.

Note that the choice of monitor point in the `stop_when_fields_decayed` routine should produce results which are independent of whether symmetries are present or not.

**Comment by oskooi** [comment]
For reference, we recently added a new FAQ on [checking convergence](https://meep.readthedocs.io/en/latest/FAQ/#checking-convergence) which provides more information and describes other possible approaches to consider.

**Comment by soamaven** [comment]
@oskooi Thanks. I had tried the field decay at a single point with the same results, namely, the 1-R and T didn't agree until very high resolution. The decay took about as long as the time allotted above, but in some cases when experimenting with different resolutions, the simulation would never reach the decay level. I  switched to using a number of periods in which I could reasonable assume the fields had decayed enough, from previous simulations.

I went further and set the simulation end criteria to be the electric power integrated in the entire volume, decayed to 1e-9 after source turn off. I saw the same behavior, where the convergence occurred for resolution of 100pts.

```
def E_power_total(dt, decay_by):
    closure = {
        'max_abs': 0,
        'cur_max': -1,
        't0': 0,
    }

    def _stop(s):

        if s.round_time() <= dt + closure['t0']:
            return False
        else:
            fabs = s.electric_energy_in_box(size=s.cell_size, center=mp.Vector3())
            closure['cur_max'] = max(closure['cur_max'], fabs)
            old_cur = closure['cur_max']
            closure['cur_max'] = 0
            closure['t0'] = s.round_time()
            closure['max_abs'] = max(closure['max_abs'], old_cur)
            if closure['max_abs'] != 0:
                fmt = "field decay(t = {}): {} / {} = {}"
                print(fmt.format(s.meep_time(), old_cur, closure['max_abs'], old_cur / closure['max_abs']))
            return old_cur <= closure['max_abs'] * decay_by
    return _stop
```
and then
```
sim.run(until_after_sources=E_power_total(5, 1e-9))
```


I guess my question at this point, is should I be trusting the Reflection spectra at lower resolutions, as its values and features seem to change less drastically with increasing resolution? 

Why would 1-R appear more precise than T? Is this surprising or expected? 

The sum indicates that the offset smoothly decreases with wavelength for 20 pts, is relatively constant for 50 pts, and at 100 pts shows a similar trend as 20 but with a smaller value. 
See below plots (sorry they are huge).

### 20 pts
![fielde_inbox_20res_crop](https://user-images.githubusercontent.com/14949210/44554549-3b288400-a6e6-11e8-9b82-5a085430d628.png)
### 50 pts
![fielde_inbox_50res_crop](https://user-images.githubusercontent.com/14949210/44554832-62338580-a6e7-11e8-95b6-40a4fcddee1d.png)

### 100 pts
![fielde_inbox_100res_crop](https://user-images.githubusercontent.com/14949210/44554551-3b288400-a6e6-11e8-9e18-3dc89f785167.png)

**Comment by oskooi** [comment]
It's difficult to say why the Fourier-transformed fields at the reflectance monitor converge more quickly than those for the transmittance. One possibility might be that the reflectance monitor is positioned closer to the source in the same contiguous media whereas the transmittance monitor is positioned further away within the high-index slab. The slab modes may be taking longer to decay away than the radiating modes in vacuum. The key is to make sure that the simulation time is sufficiently long. For the low resolution runs, did you try lowering the `decay_by` parameter in the `stop_when_fields_decayed` run function as an alternative to simply doubling the runtime?

**Comment by soamaven** [comment]
Yes, for example, for resolution of 20 pts, I allowed the volume integrated electric power to decay to 1E-15, and saw the same results as above. The fields appear to have had sufficient time to decay.

**Comment by soamaven** [comment]
As an update to this, going back to one of the other issues I pointed out, it seems that the anomalous dielectric profile at the top and bottom of the simulation region _are_ affecting the simulation at low resolutions. This is most apparent in the Ez field for the crossed grating simulated here. 

I have 3 plots, first, the T and 1-R plot I have been showing. In the middle, I have the Ez field at the end of the simulation overlayed on the dielectric function, with the various important boundaries labelled. Last (far right), I have a slice of Ez vs time _t_, with the T and 1-R as functions of _t_ (wavelength is 0.4) responses over layed on the Ez vs _t_.  Ez is normalized to the maximum value at time _t_ so that the values don't fade into white. 

The resolution here is 23, relatively low. Additionally, I did not let the simulation run until the "end" because the far left plot gets even more crowded. You may have to click on the image to view it larger.

![rt23 25](https://user-images.githubusercontent.com/14949210/45856525-c018b480-bd09-11e8-980f-c4826ee5e484.png)

The first plot shows that the T and 1-R values didn't converge, no surprises there. 

The second, gives some clues as to why. First, there is are anomalous dielectric voxels and the top and bottom of the simulation, and the fields are interacting with these. This is particularly noticeable at the bottom. The next thing I see are large kx wave vectors, so energy is basically being waveguided along the top and bottom of the interface in the simulation. Large kx and ky components are not surprising for gratings, but I am curious if this is related to [471](https://github.com/stevengj/meep/issues/471), I would need to re-read it in more detail as it has been immensely updated.

Last, the far-left plot shows that the T and 1-R will never converge on the same value. I'll guess this is just discretization error, other ideas are welcome.

 Additionally, the fields trapped in the anomalous dielectric layers at the top and bottom of the simulation _do_ appear to couple to the simulation. I don't think this is really a problem, as it is just reflection similar to what would happen when seeing the PEC at that boundary anyway, and they decay quickly inside the PML. I don't know, maybe I am making a mountain out of a mole-hill, is this intentional?

I can provide other plots/gifs/code if requested.

I think moving forward, taking the R value as "correct" for shorter, coarser simulations is the better way to go (for these grating structures at least, due to the wave-guided fields in the substrate decaying more slowly).  I will also likely need to increase the thickness of the PML and bump up the resolution. Doing these things should hopefully get me a happy compromise between speed and accuracy.

**Comment by soamaven** [comment]
Closing this. The oblique scattered waves, similar to #471, can be handled by increasing the length of the absorber layer to 1-2 lambda (PML or Absorber, `mp.Absorber()` seems to work better), decrease the `R_asymptotic` value, and/or increasing the smoothness of the such as`pml_profile=lambda u: u **5`.

The spurious index voxels at the edges come from the periodic boundary conditions, namely I did not expect the Z boundaries to _also_ be periodic in addition to the X and Y directions, rather I expected them to be metallic. I don't think this ends up mattering. One can see this explicitly by having the substrate extend beyond the lower edge of the simulation and it will show up at the top.

But I do consider it a non-critical bug, there shouldn't be any epsilon averaging in the Z direction, as even with PBC in Z, there is no discontinuous epsilon function.
